modules/scala/almond-spark/src/main/scala/org/apache/spark/sql/almondinternals/ProgressSparkListener.scala

```diff
@@ -55,6 +60,23 @@ final class ProgressSparkListener(
   def stageElem(stageId: Int): StageElem =
     elems.get(stageId)
 
+  private val ec = ExecutionContext.global
+  implicit val contextShift: ContextShift[IO] = IO.contextShift(ec)
+  implicit val timer = IO.timer(ec)
+
+  private def asyncPollUpdatesFor(elem: StageElem): Unit =
```

Comment:
It should be possible to do that with a `java.util.concurrent.ScheduledExecutorService` rather than fs2. You can [schedule](https://github.com/coursier/coursier/blob/efcfddc56213d41d37f3076160ea59952588c20c/modules/cache/jvm/src/main/scala/coursier/cache/loggers/RefreshLogger.scala#L182) calculations with it, and [cancel them](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Future.html#cancel-boolean-).

A custom `ScheduledExecutorService` could be put inside `object ProgressSparkListener { … }` for that. If it's created just like [this one](https://github.com/coursier/coursier/blob/f81c8c76fcf406b39fdf32f55769030a24889c65/modules/cache/jvm/src/main/scala/coursier/cache/internal/ThreadUtil.scala#L42), with `setKeepAliveTime` and `allowCoreThreadTimeOut` in particular, the overhead should be minimal (single thread, shutting down automatically after one minute of inactivity).

Reply:
Thanks for the pointers. WIll do a new version of it.

Reply:
@alexarchambault re-impl with java standard lib in 2nd commit